File: hashing/ransom_note.java
public Map<Character, Integer> makeCountsMap(String s)
{
    Map<Character, Integer> counts = new HashMap<>();
    for (char c : s.toCharArray())
    {
        int currentCount = counts.getOrDefault(c, 0);
        counts.put(c, currentCount + 1);
    }

    return counts;
}

public boolean canConstruct(String ransomNote, String magazine)
{
    if (ransomNote.length() > magazine.length())
    {
        return false;
    }

    // Make a counts map for the magazine
    Map<Character, Integer> magazineCounts = makeCountsMap(magazine);

    // For each character in the ransom note:
    for (char c : ransomNote.toCharArray())
    {
        // Get current count for c in magazine
        int countInMagazine = magazineCounts.getOrDefault(c, 0);

        // If there are none of c left, return false
        if (countInMagazine == 0)
        {
            return false;
        }

        // Put updated count for c back into magazineCounts
        magazineCounts.put(c, countInMagazine - 1);
    }

    // If we got this far, can successfully build note
    return true;
}

File: hashing/ransom_note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @returns {boolean}
 */
var canConstruct = function(ransomNote, magazine) {
    const makeCountsMap = s => {
        let counts = new Map();
        for (const c in s)
        {
            counts.set(c, (counts.get(c) || 0) + 1);
        }

        return counts;
    };

    if (ransomNote.length > magazine.length)
    {
        return false;
    }

    let magazineCounts = makeCountsMap(magazine);

    for (const c in ransomNote)
    {
        if (!magazineCounts.has(c))
        {
            return false;
        }

        const countInMagazine = magazineCounts.get(c);
        if (countInMagazine == 0)
        {
            return false;
        }

        magazineCounts.set(c, countInMagazine - 1);
    }

    return true;
}